download/download.go

```diff
@@ -134,6 +133,7 @@ func (d *Downloader) download(ch chan<- error, f *file, a int) {
 			return
 		}
 	}
+	defer r.Body.Close()
```

Comment:
Closing the body here won't fix the problem of holding connections. The easiest way I see to keep it recursive is to drain the connection and work with an in-memory buffer. To do so, change the lines 117-122 to:

```go
        r, err := func(p string) (io.Reader, err)
	     r, err := d.client.Get(p)
	     if err != nil {
		    log.Output(2, fmt.Sprintf("HTTP request to %s failed: %v", p, err))
		    return nil, err
	     }
             defer r.Body.Close()
             b, err:=io.ReadAll(r.Body)
             if err != nil {
		    log.Output(2, fmt.Sprintf("error reading remote file %s failed: %v", p, err))
		    return nil, err
	     }
            return bytes.NewReader(b), nil
     }()
     if err != nil {
         ch <- err
        return
    }
```

Remove this line and change line 147 to use `r` instead of r.Body. Long life to Go's io interfaces!!!!

In that way, you could safely use the recursion and be sure that you free the connection to be re-used by the next request.

Reply:
This bit `io.ReadAll(r.Body)` wasn't ideal, because reading a full body of a couple of gigabytes was taking too long and the progress bar was, therefore, frozen (and suddenly updated in big chunks). But I kept this design of wrapping the request in a function, and the logic for exp. backoff outside it… just testing once more and gonna push the changes : )

Reply:
Great. One thing that just crossed my mind: maybe removing the recursion would make things simpler? 😭

Reply:
TBH, I think wrapping that all in a function (see 899f25c) actually left the code more organized : )